share/doc/wake/basic_wake_tutorial.md

```diff
@@ -0,0 +1,660 @@
+def fttl n = n | map (\x tailstring x)
+```
+
+* Prepend and Appends every element in a list of string by "s" "m"
+`def fsml n = n | map (\x "s{x}m")`
+
+### Creating your own Data types
+* Example shows how to create a data type of our own
+```
+global data Direction =
+  North
+  South
+  East
+  West
+```
+
+* Example shows the data type Money which can be in the form of Cash or the Cheque.
+```
+global data Money =
+  Cash Integer
+  Cheque (Pair String Double)
+```
+```
+def amount  = Cash 50
+def chequeamount = Cheque (Pair "HDFC" 350.00)
+```
+
+* Try out these and record the output
+```
+def queuePut x r = foldl (\a\b b,a)x r
+def queuePut1 r = foldl (\a\b b,a) Nil r
+```
+
+* Try out these examples
+```
+def listexample1 l e = l ++ (e,Nil)
+def listexample2 l = l | head | getOrElse 0 
+def listexample3 l = l | tail
+```
+`wake -x '("abc","def",Nil) | map(explode)|map ( head)'`
+
+* Find the sum of elements in a list using foldr function. Here n refers to the deafult value which the function retursn if the list l is empty(Nil).
+`def suma l n= l | foldr(_+_)n`
+
+* Nested Function: Example
+```
+def foo n = match n 
+  0 = "toff"
+  n = bar(n-1)
+def bar n = match n
+  0 = "beerut"
+  n = foo(n-1)
+```
+
+* Finds the mean in the given list of integers
+`def mean l = l | foldr(_+_)0 | (\x x/ len l)`
+
+* Finds the median in the given list of integers
+```
+def median l = 
+  def length = if (len l %2 ==0) then ((len l) /2)-1 else ((len l)/2)
+  l | drop length | head
+```
+
+### Queues
+The front of the queue is at the right, nearest the bus stop, items are added to the left. Consider the bus queue shown, boris is at the front of the queue, ivan is last. 
+`"ivan" $$ "tanya" $$ "boris" $$ P`
+* Created a data type Queue which is nothing but a FIFO.
+```
+data Queue a = 
+  P
+  a $$ Queue a
+```
+The operations on a queue are front and remove. front returns the element at the front of the queue (without altering the queue), remove returns the rest of the queue with the first element removed. Note that both of these are strictly functions not procedures, remove does not change an existing queue it simply returns part of the queue, the original is still intact after the function call.
+
+* front returns the first element in a queue
+```
+def front = match _
+  x$$P = x$$P
+  P = P
+  x$$q = front q
+```
+
+* remove returns the rest of the queue leaving the first element in the queue
+```
+def remove = match _
+  x$$P = P
+  P = P
+  x$$q = x $$ (remove q)
+```
+
+* Converts list into a queue
+```
+def l2q = match _
+  Nil = P
+  h,t = h $$ (l2q t)
+```
+
+* Converts queue to a list
+```
+def q2l = match _
+  P = Nil
+  x$$q = x,(q2l q)
+```
+
+* Reverses the order of the queue
+```
+def doomsday = match _
+  P = P
+  q = front q $$ doomsday (remove q)
+```
+## Tree
+Tree is already implemented in wake by default.
+* Creates a Tree from a list
+`wake -x 'seq 5 |map str| listToTree (_ <=>* _)'`
+`wake -x 'listToTree (_ <=> _) (seq 20)'`
+
```

Comment:
No where in this tutorial have you explained the concept of "case classes" or whatever wake calls them and the confusing operators that get created that modify case classes. This is like 99% of the wake code that people seem to actually write at SiFive, not complicated math and map stuff, so please define those and delete a bunch of the function examples that aren't that helpful. Every example that you show should have a clear reason for its existence and thought about what you are tryign to show with that example

Reply:
Are you stating this doc should have explanation / examples of the Option ? Yes, I'd like to learn about it.

Reply:
add newlines ... two wake commands are sitting in one line when view the file.

Reply:
Done